### ISSUE TITLE
Implement breadcrumbs for glossary author node authoring

### ISSUE THREAD

**Issue by davidradl** [odpi/egeria-react-ui#30]

**Comment by davidradl** [comment]
### Existing code
Currently the url we use to show all the glossaries is 
.../<tenant>/glossary author/ glossaries 
There isa button on the uio that uses this url 
.../<tenant>/glossary author/ glossaries/add-glossary
If a user chooses a glossary, then some action buttons appear for quick terms, children, delete , edit. The urls for these are like this:

.../<tenant>/glossary author/ glossaries/edit-glossary/<glossary guid>
.../<tenant>/glossary author/ glossaries/delete-glossary/<glossary guid>
.../<tenant>/glossary author/ glossaries/quick-terms/<glossary guid>
.../<tenant>/glossary author/ glossaries/<glossary guid>/children/

Node that the children url is inconsistent with the others

Under children Terms and categories can be selected with URLs like this:
.../<tenant>/glossary author/ glossaries/<glossary guid>/children/terms
.../<tenant>/glossary author/ glossaries/<glossary guid>/children/categories

then for the edit of term 
.../<tenant>/glossary author/ glossaries/<glossary guid>/children/terms/edit-term/termguid

The above segments are defined in React routes and linked to using React Links.

### Issues with the current URL formats 

- The above url structure is not very consistent, it is sort of the rest resource way of doing urls in places but there are also actions in the path 
- and is not easily mappable onto breadcrumbs, as some of the segments do not have routes.
- a page is not fully described in the URL, for example a page that has a ticked glossary should be able to be book marked and the same page re rendered.  
- categories that have children , that have children etc, cannot be easily represented.
### Proposal

1. Amend the url formats so that they are formally resources only and following the pattern
<plural resource name> / <resource identifier> /<plural resource name> / <resource identifier> / ......
2. there is a component per segment in the path
3.  the React routes are therefore much simpler as they refer only to the path. 
4. the actions and other information to be stored in the query params. The component then needs to analyse the query params and render the appropriate component.
5. The bread crumb processing is one to one with the path with <plural resource name> / <resource identifier>  
6. The recursive category children should be able to be represented using the match and params React router. I need to work through the detail on how this will exactly work   

**I am wondering whether I should:**
1.  use something like this [https://ui.dev/react-router-v4-recursive-paths/](https://ui.dev/react-router-v4-recursive-paths/). Like this :
`<Link to={`{match.url}/${id}}>`
and the Route we render will match on the same pattern then render the same component.

`<Route path={`${match.url}/:id`} component={GlossaryAuthorComponentChooser}/>` 
The GlossaryAuthorComponentChooser then need to interrogate the path and query params and render the appropriate component.
2. Or whether the path routes should be declared and try to only recurse in the category children part, maybe use option 1 in the category part only. 

I am thinking option 2 seems preferable.  
  
Any thoughts / feedback on the above?
@planetf1 @sarbull @grahamwallis @mandy-chessell

**Comment by planetf1** [comment]
I do find the original URLs rather non-standard & very much agree with your proposal to switch to using resource names/types (nouns) in the URL path

I can't really comment further without understanding the react routing and component mapping you refer to in more detail, hopefully the others on copy can.

**Comment by sarbull** [comment]
we had the exact discussion today on odpi/egeria-ui

i wrote a good HTTP Rest Conventions best practice here [0]

i think the actions should not have different URLs, but rather one endpoint with different payload, check my rest-conventions documentation here [1]

[0] - https://github.com/sarbull/rest-conventions
[1] - https://github.com/sarbull/rest-conventions#post-apielementsidactions

**Comment by davidradl** [comment]
@sarbull thanks for responding to this. There are 2 types of URLs.
1. I follow a similar convention to your Rest API , but I do not prefix with /Api. I notice there are a number of documented conventions on the web, for example
-  https://docs.microsoft.com/en-us/azure/architecture/best-practices/api-design#:~:text=REST%20is%20independent%20of%20any,not%20necessarily%20tied%20to%20HTTP.&text=For%20REST%20APIs%20built%20on,use%20a%20stateless%20request%20model.
- https://cloud.google.com/apis/design

 I am unsure how what you have documented differs from these.

 I am aware of 2 approaches - one where reference uris are included in payload responses, in this case the rest API can be very chatty. I have worked on code that took this approach but mostly I see it not being used for user interfaces. The other approach is to return bigger payloads representing everything that is needed by the UI. I have used the less chatty API approach.  

2. The Url displayed in the browser. The complete discussion above refers to these URLs - not Rest APIs. I am following a noun orientated approach (as mentioned by Nigel), so I need a url to say represent a create page - that will render a create component using react routing; I am putting the fact it is a create page in query params.
